util/tls/tls_socket.cc

```diff
@@ -220,10 +226,16 @@ io::Result<size_t> TlsSocket::RecvMsg(const msghdr& msg, int flags) {
       continue;
     }
 
-    if (read_total > 0 && op_val == Engine::NEED_READ_AND_MAYBE_WRITE) {
-      // If we have read some data, we should not block on further reads.
-      // TODO: for async reads though we could issue a read request since we know the engine
-      // buffer is empty.
+    if ((read_total > 0) && (op_val == Engine::NEED_READ_AND_MAYBE_WRITE)) {
+      // After reading some data, the TLS engine may have generated protocol output (e.g., handshake
```

Comment:
a valid suggestion but why not handle it via HandleOp? instead of duplicating the code here, maybe we can call HandleOp before the condition at line 229? 

also again - I do not think it is related to TLS async and could be done in a preliminary PR with all the improvements/changes you made that are unrelated to TLS async

Reply:
I have added the above due to breakpoint debugging, just saw that there is data  to be flushed (multiple times).

Thanks for the suggestion, I have encapsulated as suggested the logic in both HandleOp and HandleOpAsync, and I'll upload the change in the next PR, outside the TLS PR.